# Acknowledgements {-}

I would like to thank LIFE-HUB.CSIC Network for providing financial support to my involvement in this project. Also, thanks to Doctors Rosa Fernández (IBE) and Ana Riesgo (MNCN) for kindly sharing their transcriptomic data with us.

I would also like to express my gratitude to Professor Santiago F. Elena for counting on me for this project and accepting me once again in his laboratory, apart from reviewing this document. I would also like to extend my thanks to everyone in the Evolutionary Systems Virology group of the Institute for Integrative Systems Biology (I$^2$SysBio) for sharing insightful comments and appreciations about this project with me.

The development of bioinformatic methods would not have been possible without the help of Doctor Anamarija Butković, to whom I hereby express my deepest gratitude. Also, thanks for the first revisions of this document.

I am also thankful to Professor Francisco J. Santonja for his supervision over the statistical methodologies and his recommendations on bibliography, especially at the starting point of the project.

Thanks to my parents and grandmother for supporting me in all the ways possible through all my years of study. *Vos vullc*.


# Abstract {-}


Current knowledge on the viruses that coexist with invertebrates is very limited and biased towards clades of interest. The application of (meta)viromics pipelines over genomics or transcriptomics sequencing data allows for a large-scale identification of the viruses present in the sequenced sample. Metaviromics has been applied recently to some phyla of invertebrates in order to characterize their viromes, but viromes of certain phyla have been left unexplored to present date. In this study, publicly accessible transcriptomics data of very diverse invertebrate species has been used to discover how phylum and habitat affiliations shape their associated community of RNA and DNA viruses.

To achieve this goal, the development and selection of a bioinformatics pipeline for the analysis of viromes in transcriptomics data was conducted. The resulting virome compositions at a family level are set to be analyzed with compositional data analysis methods. Centered logratio (*clr*) transformations are used to transform the compositions from closed data to open data so as to apply statistical multivariate methods such as PCA. It was found that invertebrate phylum can explain around $30\%$ of the total variance in the estimated virome compositions, whereas habitat did not have such a significant impact. Overall, the identified viral families and the abundance of their transcripts were surprisingly stable among the different phyla and the viromes were dominated by transcripts of DNA viruses.


# Introduction

The aim of this study is to describe the diversity of viruses associated with invertebrates based on high-throughput sequencing data analysis. Before focusing on the details of the study, let us introduce viruses, invertebrates and the basis of massive sequencing.


## Viruses

### Historical insights on the definition of a virus

The word ‘virus’ has been used for hundreds of years as a broad term to refer to the causal agents of infectious diseases. In the late 19th century, the first hints of pathogens smaller than bacteria were found, just a few years after the works of Louis Pasteur and Robert Koch that led to a revolution in microbiology and to the proposal of the germ theory, which originally stated that microorganisms are the causal agents of infectious diseases. In this context, different authors found that the pathogens involved in infectious diseases they were studying failed to satisfy the praised criteria of Koch’s postulates, because of the inability of viruses to replicate *ex vivo*.

In the field of plant pathology, Martinus Beijerinck’s findings regarding his research on tobacco mosaic disease published in 1898 can be considered the seed for the later development of virology, as he was convinced he was dealing with an infectious agent different in nature to microbes [@Bos1999]. Among his conclusions about the newly discovered pathogen, i.e. the tobacco mosaic virus, distinctive characteristics of viruses stand out: *(i)* the infectious agent is smaller than any known bacteria, as it cannot be retained with a Chamberland filter, which were thought to not let bacteria pass through at the time, *(ii)* it is unable to grow in artificial solid media, but isolates from diseased tissues remain infectious even after a few months, *(iii)* it relies on the host’s active metabolism to reproduce.

Parallelly with Beijernick’s discoveries, the results published by Friedrich Loeffler and Paul Frosch on the causal agent of foot-and-mouth disease in cattle deserve a special mention. However, they placed the emphasis on the filterability of this pathogen, now known to be a virus of the family *Picornaviridae* [@Jamal2013]. Its undetectable size by then existing microscopy and its apparent inability to be independently cultivated *in vitro* given the available culture media at the time, led them to the conclusion that this new submicroscopic germ could not be properly studied [@Mahy2005]. Although they did not categorically refute the applicability of Koch’s postulates to this agent, they were right to consider the corpuscular nature of the pathogen, instead of a liquid or soluble infectious principle as Beijerinck proposed [@Bos1999]. Thus, the first cases of disease in plants and animals induced by so-called ‘filterable viruses’ were discovered.

The systematic approach of bacteriology to the identification of pathogens led to the finding of numerous diseases that were caused by ‘filterable viruses’ in the following years, with much of the attention drawn merely by the size of these agents. However, this alone is not a definitory feature of viruses and it caused a long-lasting confusion in the field. This was exacerbated by the discovery of the agent of contagious bovine pleuropneumonia, a mycoplasma of the bacteria family Mollicutes [@E.Nocard1990]. These small prokaryotes were able to pass through Chamberland filters because of their submicroscopic diameter of around $0.2$ $\mu m$, but showed growth on solid media under specific conditions. It is currently known that virions, i.e. individual viral particles, of giant viruses of nucleo-cytoplasmic large DNA virus (NCLDV) families can be larger and measure up to $0.6$ $\mu m$ [@Brandes2019].
Apart from that, this criterion alone would make viruses indistinguishable from other potentially pathogenic submicroscopic and subcellular entities, such as prions, viroids and plasmids. Future technical and theoretical developments in biochemistry, molecular biology, genetics and ultimately biotechnology would shed light on the subject of virology throughout the 20th century.

In the 1930s, William Elford estimated the size of different viruses, small bacteria and reference materials as egg albumin thanks to the advancements in ultrafiltration [@Elford1932]. As a result, the overlap between the size range of viral particles with the size of the smallest bacteria in the upper end and that of ‘inanimate’ molecules in the lower end became a known fact. This placed viruses in a position between the subjects of study of chemistry and biology; a place where "a sharp line dividing living from nonliving things could not be drawn", as Wendell Stanley stated in his Nobel prize lecture in 1946 [@Stanley1946]. In 1937, the Nobel laureate in chemistry observed the crystallized rod-shaped virions of tobacco mosaic virus for the first time, identifying the particles as nucleoproteins [@Stanley1938].

In 1957, André Lwoff presumably proposed the first modern definition of viruses mixing both the intrinsic characteristics of viruses and the features of other biological entities, especially cell-based organisms, that lack in viruses. Thus a virus was defined as a potentially pathogenic nucleoproteinic entity able to replicate by taking its genetic information, i.e. nucleic acid (DNA or RNA), as a template, in contrast to the binary fission that occurs in cells after a certain degree of growth is reached. Besides, viruses do not encode complete metabolic systems to use the energy contained in the covalent bonds of organic matter and therefore are energetically dependent on their host cells [@Lwoff1957]. In 1961, François Jacob & Élie Wollman added interesting aspects to this definition such as the enclosure of the nucleic acid in a protein coat and the ability of viruses "to undergo changes in their hereditary properties", which implies that viruses are subject to selection and therefore evolution [@Jacob1961].

However, these views are heavily biased by the extracellular phase of viruses, where the individuality of viruses appears intuitive in the form of encapsidated genetic elements that constitute virions. Today, this "virus/virion" paradigm is qualified, if not disregarded [@Forterre2016], in order to define viruses from a theoretical viewpoint, as it conceals the complexity of host-virus interactions in the intracellular phase. In this situation, the individuality of viral particles is lost as the components of virions disperse inside the cell and integration in the host genome can occur. Along with the ongoing question of viruses being alive or not, the definition of viruses is still not an objectively solved issue by virologists and numerous objections arise to any given concise definition given the new discoveries, from capsidless RNA viruses [@Dolja2012;@Zhang2016] to giant DNA viruses encoding parts of metabolic pathways [@Schvarcz2018], metabolically inert cell-based life forms as endospores of Gram-positive bacteria [@Galperin2012] or obligate intracellular parasitic bacteria [@McClure2017].


Outside the "virus/virion" paradigm, it is proposed that viruses should be defined as the biological process that encompasses the whole viral reproduction cycle, including virions and intracellular states of viruses such as viral factories or integrated proviruses [@Forterre2016]. Moreover, other authors embrace the replicator paradigm [@Dawkins1976] in order to define viruses, placing them in a continuum of nucleic acid-based replicators, along with cellular chromosomes, mitochondrial DNA, plasmids and viroids. Viruses can be distinguished from other replicators in different axes: replicative autonomy, reproductive strategy (selfishness vs. cooperativeness), resource production and the nature of the vehicle used for resource obtention and progeny spread, i.e. virions in the case of viruses [@Koonin2016].

### Quick overview on the origin, evolution and ecology of viruses

In terms of abundance and genetic diversity, viruses are arguably the most evolutionarily successful biological entities on our planet, even if an important part of the virosphere still remains unknown [@Wasik2013]. Viruses, especially bacteriophages, i.e. those that infect bacteria, outnumber cellular life-forms in most environments, e.g. in marine water or sediments [@Edwards2005]. It is their replicative strategy, generally characterized by their compact genomes, high progeny numbers and high mutation rates, especially in RNA and single-strand DNA (ssDNA) viruses [@Belshaw2007; @Sanjuan2010], that makes viruses particularly capable of adaptation to environmental changes. Although it is important to note that this adaptation is not independent from the host and in this sense, it is effectively the holobiont formed by the virus and the host that adapts [@Agudelo-Romero2008].


Viruses infect organisms from the three domains of life: Archaea, Bacteria and Eukarya, the latter including plants, vertebrates and invertebrates. Although their origin is still a subject of dispute by evolutionary virologists, these biological entities of relatively low complexity are thought to be approximately as ancient as cellular life [@Holmes2011; @Koonin2006]. However, even if some homologies exist between viruses of these very distant hosts, it is accepted that viruses do not have a common ancestor and have a polyphyletic origin instead, as there is not a single universally viral gene. A chimeric origin of viral genomes has recently been proposed, in which the replicative part of the genome descends from an ancient/primordial genetic pool, especially in RNA viruses, whereas the coat protein-encoding part has had different origins and the structurally limited set of extant capsid folds may be a product of convergent evolution. This theory places viruses in an evolutionary network rather than a tree and highlights recombination and horizontal gene transfer as major drives in virus evolution [@Krupovic2019].


Besides, viruses have had a major impact shaping the evolution of the entire biosphere from the very start of cellular life forms [@Forterre2006]. Apart from the complexity of host-virus interactions and host specificities, modern genomes of eukaryotic organisms show the intimate history of host-virus coevolution in the form of integrated endogenous viral elements (EVEs) [@Feschotte2012].

Contrary to popular belief, the majority of viruses do not form pathogenic or antagonistic symbiotic relationships with their hosts, i.e. most viruses do not cause disease on their hosts. Moreover, most often viruses adopt commensal lifestyles, in which they benefit from the hosts but do not cause significant harm to them. Also numerous mutualistic relationships have been described in the past years, where viral infection is in fact advantageous to the host [@Roossinck2017; @Gonzalez2021]. Given the complexity of these interactions, the outcome of infection in terms of host phenotype alteration can further depend on environmental factors [@Rodriguez2008; @Gonzalez2021].



## Invertebrates

Animals, also called metazoans, are multicellular eukaryotic organisms characterized by being motile to a greater or lesser degree [@Arnellos2019] and having heterotrophic lifestyles [@Mills2017], i.e. they rely on other organisms as sources of organic carbon. Animals as a whole form a monophyletic group with a common ancestor dated around 800 million years ago, as shown by recent molecular clock studies [@Erwin2015]. As it happens with the entire tree of life in general, the phylogeny of animals in particular is still not completely resolved and is undergoing major reorganizations [@Giribet2019].

Invertebrates constitute a taxonomically informal, paraphyletic grouping of animals that conveniently excludes vertebrates, i.e. those animals that develop a vertebral column, from the rest of metazoans. Although other ancient traits, e.g. the development of a notochord or symmetry patterns in the embryogenesis, divide animals in monophyletic groups, the term is still used for practical reasons as it fits well with an anthropocentric view of the world.


Moreover, invertebrates harbor a huge biodiversity and make up roughly 95 \% of the total estimated number of described animal species, of which 83 \% belong to arthropods alone, making them by far the most successful animal phylum^[Phylum (*plural* phyla): a rank located below kingdom (e.g. Animalia) and above class (e.g. Mammalia) in the taxonomic hierarchy. It is equivalent to the 'division' rank that is used in plants and fungi.] in terms of diversity [@Zhang2013]. Considering how underexplored invertebrates are in relation to vertebrates, this proportion can be expected to grow even higher as previously unknown invertebrate species are discovered.


Within invertebrates, the exploration of different phyla has been highly asymmetrical, focusing mainly in species relevant to human health or with direct socioeconomic implications, e.g. insects or nematodes. This results in a spatial and taxonomic bias that has historically neglected invertebrates of many other phyla, with much emphasis on marine species. However, efforts are being made in order to describe the diversity of understudied invertebrates in all types of habitats, from marine to terrestrial species [@Lopez2014].

The biodiversity and abundance of invertebrates reflects their critical role in the functioning of ecosystems, e.g. as agents of pollination or decomposers of dead wood and leaf litter. In recent years, the "silent extinction" that many invertebrate species are facing, apart from an overall insect biomass decline in some areas, has gained public awareness  [@Hallmann2017; @Eisenhauer2019]. The cascade effect of a large-scale reduction of invertebrate biodiversity raises serious concerns about the magnitude of its ecological and socioeconomic consequences. Therefore, a better understanding of invertebrates would make it possible to identify the causes of the extinction and apply monitoring and conservation measures.

One fundamental source of such understanding is indeed obtained at a molecular level by nucleic acid sequencing, since visual identification of species is not always viable [@Chen2021]. The bias in the study of invertebrates is further exacerbated by the lack of widespread genetic information in undersampled phyla, although again this is a well-known problem that is currently being adressed by zoologists [@Laumer2019]. Furthermore, sequencing of ecological invertebrate samples can further capture information about the microbes and viruses that coexist with them. This is not only relevant to describe the interaction of invertebrates with other organisms of their environment, but direct applications include the surveillance and prevention of disease outbreaks in humans, livestock [@Modha2019], fisheries [@Ferguson2010] and plant crops [@Kondo2020]. This is because many invertebrates serve as vectors of infectious diseases, transferring pathogens from one host to another.

The threat of vector-borne diseases has grown in recent times as a consequence of global trade and climate change, among other factors [@Baker2021]. For example, mosquitoes of invasive species *Aedes albopictus* serve as vectors of tropical diseases such as dengue and chikungunya, which have already shown autochthonous transmission in Europe [@Gossner2022;@Tjaden2021]. With the rise of mean temperatures in the last decades, these mosquitoes have established in southern Europe and are expanding their range to higher latitudes. Viruses can form complex interactions with their vectors, for instance, replication within the invertebrate vector is common in animal and even plant viruses [@He2020]. Furthermore, some viruses manipulate their vectors as a strategy to promote their spread [@Eigenbrode2018;@Jayasinghe2021]

## Introduction to metaviromics

### What are omics sciences?

The advancements and cost reduction in sequencing and instrumental chemistry in the last decades have led to the development of the so-called omics sciences, and consequently a revolution in the generation of data in modern biological science. Omics sciences consist of high-throughput, massive analysis and profiling of a pool of biochemical molecules of choice within a particular biological sample or system. Therefore, omics involve a holistic approach for the understanding of a biological problem with the possibility to obtain and interpret both qualitative and quantitative information. In many cases, they serve as a hypothesis-generating method when there is little known about the process or issue in question [@Vailati-Riboni2017].

Depending on the class or nature of these molecules, a different omics field is defined. Genomics places its focus on DNA molecules within a cell, tissue or organism, while transcriptomics involves the analysis of RNA molecules, typically focusing on mRNA, small RNA and non-coding RNA transcripts, which form the transcriptome. Other fields such as proteomics and metabolomics see their target molecules in proteins and small metabolites, respectively.

The subjects of different omics, the *omes*, are closely connected in biological systems in line with the central dogma of molecular biology: genes are transcribed into RNA molecules, some of which are translated into proteins, many of which affect metabolite levels. Furthermore, the combination of different omics datasets in the multi-omics framework can provide deeper insights for the understanding of a biological process [@Krassowski2020].


```{=latex}
\begin{figure}[!htbp]
```
```{r sra-tb, echo=FALSE, out.width='80%', fig.asp=.75, fig.align='center'}
library(ggplot2)
library(ggfortify)
library(scales)
sradat <- read.csv("data/sra_stat.csv", header = T)
obj <- as.Date(sradat$date, format = "%m/%d/%Y", origin = sradat$date[1])

breaks = 10^(-10:10)
minorbreaks = rep(1:9, 21)*(10^rep(-10:10, each = 9))
ggplot(sradat, aes(x=obj, y=bytes/1099511627776)) +
  geom_point() +
  scale_x_date(breaks = seq(as.Date("2000-01-01"), as.Date("2050-12-31"), by="1 year"), 
               labels=date_format("%Y"), minor_breaks = as.Date("01/01/2000")) +
  scale_y_log10(breaks = breaks, minor_breaks = minorbreaks,
                labels = trans_format("log10", math_format(10^.x))) +
  theme_bw() + 
  annotation_logticks(sides = "l") +
  xlab("Year") + 
  ylab("Total size of sequence files stored (Tb)")

```
```{=latex}
\caption{Total size (Tb) of sequence files stored in NCBI's SRA database. \label{fig:sra-tb}}
The SRA database hosts plenty of publicly accessible results of nucleic acid high-throughput sequencing experiments shared by the National Center for Biotechnology Information (NCBI), the European Bioinformatics Institute (EBI) and the DNA Data Bank of Japan (DDBJ). Data retrieved from https://www.ncbi.nlm.nih.gov/sra/docs/sragrowth/ on 23th march 2022.
\end{figure}
```


Typically, the number of variables $p$ greatly outnumbers that of observations $n$ in omics datasets, especially in those of genomics and transcriptomics. This imbalanced high-dimensionality poses several challenges for statistical analysis, such as variable selection, studying interaction effects or the trade-off between false positives and false negatives when testing thousands of hypotheses [@Korthauer2019; @Yamada2020]. Also, from a computational viewpoint, storing and processing of omics raw data requires the development of efficient methods, algorithms and pipelines, which has been an important task in the field of bioinformatics [@Berger2013] as the volume of omics data generated kept growing every year (Figure \@ref(fig:sra-tb)).

Because of the high amount of information contained within a single omics dataset, it is virtually impossible to consider it exhausted after the analysis by its original authors. The application of different methods on a single omics dataset yields different results [@Yamada2020]. In this sense, it can be assumed that two experimental steps exist: before and after raw data generation, the former *ex silico* and the latter *in silico*. Therefore, public availability and proper metadata annotation of omics data is not only seen as a good practice in terms of reproducibility, but also for reusability [@Perez-Riverol2019].

### (Meta)transcriptomics in the era of RNA-Seq

The development of deep-sequencing technologies, alternatively called next-generation sequencing (NGS) technologies, has effectively widened the range of possibilities within genomics as well as transcriptomics. In the second half of the 2000s, the irruption of RNA-Seq caused a qualitative leap with respect to Sanger sequencing and hybridization-based methods in many aspects. Compared to microarrays, RNA-Seq *(i)* does not rely on previously studied genomes and is therefore suitable for *de novo* approaches on non-model organisms, *(ii)* has a higher sensitivity and a much larger dynamic range of RNA molecule abundance levels and *(iii)* its results are more reproducible, allowing for robust cross-experimental comparisons [@Wang2009].

To date, second generation sequencing data continues to dominate public repositories of transcriptomic data, although third generation approaches are thought to establish as a standard technology in the near future [@Athanasopoulou2021]. In short, a typical *shotgun* style second generation RNA-seq workflow involves first the isolation of total RNA from a biological sample, followed by an optional step of RNA species (mRNA, rRNA, miRNA, etc.) selection. Then, the remaining population of RNA molecules is subject to fragmentation into shorter sequences of up to 600 bases, due to limitations of second generation sequencing platforms. Afterwards, RNA molecules undergo conversion to cDNA catalyzed by the enzyme reverse transcriptase, and so a cDNA library is constructed. In the process, a series of adapters and barcode identifiers are attached to the ends of the RNA/cDNA fragments by ligation and/or amplification. Each cDNA fragment is sequenced from one end (single-end reads) or both ends (paired-end reads), with a read length of around 100 bp [@Hrdlickova2017]. Once the RNA-Seq read data is formatted in a text file, the bioinformatic procedure begins. Here, a common first goal is to reconstruct the RNA molecules existing in the original sample by aligning overlapping regions between the reads. Finally, the abundance of each RNA molecule can be inferred from the read count assigned to each assembled sequence after a length correction. The detection sensitivity and the accuracy of abundance estimation depend on the read depth, which must be chosen according to the prior knowledge of the biological sample and the aim of the experiment [@Kukurba2015].

One fundamental problem of second generation sequencing is the assembly of ambiguous reads into 'chimeric' contigs^[A contig is a continuous sequence produced by alignment of overlapping reads. Contigs try to reproduce the original molecules from which a particular set of fragments was derived.], as one single read can be mapped into homologous regions of physically different molecules. The probability of a read being ambiguous decreases as the read length increases. Therefore, the sequencing of full-length cDNA (FL-cDNA) with third generation sequencing platforms is indeed promising, as this technology also accounts for mRNA splicing variants. However, the higher cost per base and the higher base calling error rates continue to hinder its rise as the most popular technology for transcriptome profiling [@Krizanovic2018].

In many cases, the motivation of transcriptomics studies is to find differentially expressed genes between conditions, focusing on the abundance of nuclear genome-encoded mRNAs. However, many other applications arise considering the diverse origin of RNA molecules that can be found within a biological sample. The prefix *meta-* is added to genomics and transcriptomics when taking into consideration the microbial DNA and RNA content, respectively. Both provide different perspectives for the estimation of the community composition of a microbiome associated to a tissue, organism or simply to an environmental sample, e.g. soil, freshwater or seawater. While in the 20th century microbiologists approached the study of microbial community and function by isolating and culturing individual species in a reductionist manner, the parallel analysis of the different and potentially unculturable components of a microbial community through metagenomics and metatranscriptomics has revolutionized the understanding of their role in geochemical processes and the biosphere. Apart from that, these new approaches have led to an unprecedented description of microbial phylogenetic diversity  [@Mardanov2018].

The choice of DNA or RNA as the source of information is the fundamental difference between metagenomics and metatranscriptomics and affects the conclusions that can be drawn from the analysis. In metagenomics, usually a phylogenetic marker like 16S rRNA gene is surveyed for classification of individuals into operational taxonomic units (OTUs). This prokaryotic gene is highly-conserved yet polymorphic enough to show differences between species, but also within species in some cases [@Pei2010]. Then, the estimated abundance of each OTU is proportional to the number of 16S rRNA of that particular OTU found within the sample, which serves as an indirect measure of the number of genomes. Although this gene alone could be targeted by PCR^[Polymerase Chain Reaction (PCR) is a method used in molecular biology to replicate the DNA sequence between two known regions in order to obtain a large number of copies of that sequence (amplification). In NGS, fragments are amplified thanks to the adapters that flank them and whose sequence is determined by design.] amplification and amplicon sequencing, the shotgun sequencing alternative is becoming more popular for its sensitivity [@Durazzi2021], lack of bias [@Brooks2015] and presence of genomic information that can complement 16S rRNA analysis. However, in regards to function, the presence of different genes provides information about potential function, but does not serve as an evidence for effective gene expression.

Meanwhile, metatranscriptomics is more suitable for functional annotation, because RNA molecules are generally the product of transcription and therefore show activity of the genes of a community member. However, it can be argued that mRNA presence alone does not strictly result in gene expression, i.e. translation into functional protein [@Vogel2012]. While it is also possible to perform taxonomic classification of reads or contigs in metatranscriptomic datasets similarly to metagenomics, the dynamics of RNA molecule population allow for the longitudinal study of the activity of community components over time and in response to changes in the environment [@Shakya2019].

### Virome profiling with metatranscriptomics

Alongside microorganisms, DNA and RNA pools of biological samples can also contain molecules of viral origin. Therefore, metagenomics and metatranscriptomics approaches can be directed towards the characterization of viral communities and identification of novel viruses. Again, this method for virome profiling is less biased and more powerful than culture-dependent methods [@Mokili2012]. The lesser-used term 'metavirome' is employed in the sense of a sum of viromes to stress the diversity of hosts that viruses have within a particular sample, especially in environmental ones. However, the distinction is rather vague and viruses hosted by different organisms that form part of a symbiosis [@Garcia-Lopez2019] or an ecosystem [@Kim2017] are often referred to as part of the virome.

There are several key differences between microbiome and virome analyses. First, there is no phylogenetic marker similar to 16S rRNA that is applicable to all classes of viruses [@Rohwer2002] (see \@ref(quick-overview-on-the-origin-evolution-and-ecology-of-viruses)), although  genes encoding the RNA-dependent RNA polymerase (RdRp) and  reverse transcriptase (RT) are highly conserved in RNA viruses and reverse-transcribing RNA viruses, respectively [@Koonin2009]. Second, the election of DNA or RNA as a source of information has further implications because of the different types of viral genomes. DNA sequencing can provide information about DNA virus genomes and intermediate forms of reverse-transcribing viruses and proviruses, while metatranscriptomics can survey the whole population of RNA molecules, including transcripts of transcriptionally active DNA and RNA viruses and genomes of RNA viruses.

*Metaviromics* has already reshaped our knowledge on the ecology and evolution of viruses [@Koonin2018]. For example, (a) the diversity, community dynamics and potential role in the carbon cycling of RNA viruses in soil has recently been addressed [@Starr2019], (b) longitudinal studies have shown the individuality and stability of human gut virome, in line with what was known about gut microbiome [@Shkoporov2019], (c) the discovery of the ubiquitous presence of RNA viruses in marine environments, where they were thought to be scarce [@Culley2018] and d) metaviromics has allowed for the identification of widespread horizontal gene transfer between viruses of different Baltimore classes and virus transfer between plants, fungi and animals, especially invertebrates, emphasizing the network-shaped nature of RNA virus evolution [@Wolf2018; @Dolja2018].

## State of the art in invertebrate metaviromics

In regards to invertebrate metaviromics, many recent studies focus on applying the systematic framework of metaviromics to a particular taxon of interest in order to discover novel viruses and describe the diversity of its virome. It is important to note that in most cases, the aggregate of viral sequences identified by deep-sequencing includes viruses that infect different parts of the holobiont with the invertebrate as the host. Furthermore, depending on the sampling conditions, contamination with viruses outside the holobiont can be found, e.g. viruses of plants, fungi or bacteria. Therefore, it is not a target of metaviromics *per se* to define the host range of the identified viruses, but to identify the presence and abundance of the different assigned viruses. If a found virus shows significant sequence similarity with a known virus, it can be used to deduce its original host.

Anthropocentrically important species draw most of the attention in invertebrate metaviromics. The RNA virome of hematophagous arthropods like ticks [@Meng2019] and biting midges [@Modha2019] has been recently characterized in an effort to evaluate the risk of zoonotic disease emergence in humans or livestock. At the same time, the RNA viromes of vectors of plant diseases like aphids [@Kondo2020] and whiteflies [@Huang2021] have been profiled. Thus identifying the dynamics and effects of different virome compositions on the host phenotype, consequently leading to the proposal on the use of host disease-inducing viruses as biocontrol agents [@Feng2017]. Apart from arthropods, metaviromics have been applied in echinoderms to examine the cause of wasting disease in starfish based on differences in virome composition [@Hewson2020]. Also, the DNA virome associated with the Porifera species *Haliclona fulva*, a sponge holobiont with low microbial abundance, has been recently studied, showing an unexpected low abundance and diversity of bacteriophage families [@Garcia-Bonilla2021]. As for cnidarians, the whole DNA and RNA virome for a sea anemone model species, using publicly available RNA-Seq data, showed a core of viral families found across all samples and how the virome could change depending on the association with a symbiotic alga [@Bruwer2018]. Cnidarians have historically attracted much attention in invertebrate metaviromics, not only because of their ancient divergence during the evolution of metazoans and their richness in symbiotic relationships, but also because of the environmental importance of coral reef ecosystems [@VegaThurber2008].

The examples of invertebrate metaviromics studies presented above focus on one or a narrow spectrum of related species. Other authors have focused on studying the virome profiles across higher taxonomic ranks, identifying the differences that can occur within and between phyla. In 2019, a study was published surveying the negative single strand RNA virome of more than one thousand insect species using amino acid profile hidden Markov models of the RdRp gene on transcriptomic datasets, thus identifying novel viruses and significant virus-insect co-evolution [@Kafer2019]. As for Porifera, a recent publication revealed that the DNA virome of 15 sponge species is dominated by bacteriophages of the order *Caudovirales* and NCLDVs of eukaryotes , also showing how the virome composition can change depending on the geographical location and nutritional strategy of sponges with PERMANOVA analysis [@Pascelli2020]. A study published in 2021 comparing the viromes of marine Annelida, Arthropoda and Mollusca led to similar conclusions, confirming that the geographical region has an effect on virome composition [@Zhang2021]. In 2016, a comparative analysis of the DNA viromes of four different groups of marine invertebrates (Arthropoda, Cnidaria, Echinodermata and Urochordata) acknowledged the disparities of the viromes within arthropods and the similarities between the viromes of marine arthropods and echinoderms, among other findings [@Gudenkauf2016].

Fewer studies in metaviromics have tried to assess the diversity of the viromes across a higher number of invertebrate phyla. In 2016, Shi et al. presented a pioneering work on the global RNA virosphere of invertebrates, sampling metatranscriptomic data of 220 aquatic and terrestrial species of eight invertebrate phyla. They found novel viruses, proposed new viral families, filled major gaps in the RdRp phylogeny and evaluated the degree of virus-host co-divergence and horizontal virus transfer in the history of RNA virus families of invertebrates [@Shi2016]. A related publication in 2019 surveyed the community of transcriptionally active DNA viruses in the same set of species from metatranscriptomic data [@Porter2019]. Both studies show an unprecedented viral diversity in the invertebrate-associated virosphere and stress the difficulty of host assignment to metaviromics-identified viruses and the bulk of viruses that remain unknown. The identification of new viruses is limited and biased by sequence alignment-based taxonomic classification, as the viruses more distantly related to those available in public databases will remain preferentially hidden in high-throughput analyses. Earlier this year, a review was published that discussed the state of the current knowledge on the animal virosphere, proposing the sampling of poorly explored invertebrate phyla and animals of diverse habitats as a priority for the near future in order to better understand animal virome diversity and evolution [@Harvey2022].

## Perspectives of the present work

In this study we propose a large-scale application of a metaviromics workflow on more than a thousand transcriptomic datasets of more than 30 invertebrate phyla, all publicly available in the NCBI SRA database [@Leinonen2011], also including the subphylum of chordates formerly known as Urochordata (now Tunicata). The project will cover organisms belonging to invertebrate phyla with a lack of previous virome analyses such as Kinorhyncha, Placozoa, Priapulida or Tardigrada, just to name a few. Besides, this study will also serve as an example of meta-analysis of transcriptomic data from the SRA database, discussing the state of the available metadata in the context of data reusability.

One of the main goals is to discover the variability of the RNA and DNA virome composition in terms of viral families, within and between as many phyla of invertebrates as possible. Additionally, the effect of the habitat of the host species on the viromes will be evaluated, including terrestrial, marine and freshwater samples, among others. An important point of the study will address the consistency of the results related to species with previously explored viromes in order to validate the bioinformatics pipeline employed.


